### DISCUSSION TITLE
Functional programming optimizations

### DISCUSSION THREAD

**Discussion by MackNZ** [parasail-lang/parasail#3] [Q&A] [answered]
I'm interested in writing a frontend to ParaSail where the target language would be written in a functional style, but ParaSail doesn't look as though it's typically written functionally. What types of optimizations does the backend perform for functional style code? Is tail call optimization performed?

**Comment by sttaft** [comment, marked as answer]
Interesting question!  As you presumably have seen, there are currently five front ends for ParaSail, each with its own subdirectory where the parser and lexer live:
-  Java-like: Javallel -- parasail/javallel_parser
-  Python-like: Parython -- parasail/parython_parser
-  SPARK-like: Sparkel -- parasail/sparkel_parser
-  Ada 2022: Ada2x  -- parasail/ada202x_parser
-  and of course, ParaSail -- parasail/parser
So writing a new front end is relatively straightforward, by copying the *.y and *.l files from one of the above and editing the grammars as appropriate.

But in fact, you can already write largely functional programs in ParaSail, because you can insert a `const X := F(Y)` or a `ref P => List_Head.Next` anywhere in a sequence of statements, and there is a "for" loop structure that gives you C-like iteration in a functional way, both for simple numeric sequences and for iterating over linked structures:
```
   for X := 0 then X+1 while X < 10 loop
       ... // X is not a variable here
   end loop
```
or
```
    for P => List_Head then P.Next while P not null loop
       ... // P is a reference to a list element
    end loop
```
I would suggest you try using ParaSail while avoiding use of `var` and see how far you can get before designing your own front end.  You might find that ParaSail already has most of what you want, with only a few new constructs needed.  A good way to design new constructs is to create examples on "paper" before hand.  When I was designing ParaSail, I wrote some significant programs before I started implementing, which meant I had some ready-made test programs once something began to work.

As far as optimization, I would wait on that until you have something actually running.  If you choose to take advantage of the ParaSail VM => LLVM compiler, then you will have to see what sort of tail recursion, etc., is supported by LLVM.  My guess is that it will have what you need once you get to that point, though I personally haven't looked.

**Reply by MackNZ** [reply]
Thanks!

**Reply by sttaft** [reply]
By the way, I checked, and LLVM does have support for tail recursion.  It seems you just put the keyword "tail" at an appropriate place in the LLVM call statement, and it will try to use tail recursion rather than starting a completely new stack frame.

**Reply by ec-c** [reply]
> * Java-like: Javallel -- parasail/javallel_parser
> * Python-like: Parython -- parasail/parython_parser
> * SPARK-like: Sparkel -- parasail/sparkel_parser
> * Ada 2022: Ada2x  -- parasail/ada202x_parser
> * and of course, ParaSail -- parasail/parser
>   So writing a new front end is relatively straightforward, by copying the *.y and *.l files from one of the above and editing the grammars as appropriate.
> 

Sorry to hijack this discussion, but I was wondering whether there are any differences in features between the various front ends and if one is recommended.

**Reply by sttaft** [reply]
I would recommend using the ParaSail front end.  They all share 95% of the interpreter and compiler, but I only regularly test the ParaSail one.  Also, the ParaSail one has a more complete run-time library ("aaa.psi").  The other run-time libraries were copied and translated from ParaSail several years ago, and so have fewer features (and probably more bugs ...).
